internal/keepers/factory_test.go

```diff
@@ -17,70 +18,208 @@ import (
 )
 
 func TestNewReportingPluginFactory(t *testing.T) {
-	f := NewReportingPluginFactory(
-		nil,
-		nil,
-		nil,
-		nil,
-		nil,
-		ReportingFactoryConfig{},
-	)
-	assert.NotNil(t, f)
+	t.Run("a new reporting plugin factory is created without dependencies", func(t *testing.T) {
+		f := NewReportingPluginFactory(
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			ReportingFactoryConfig{},
+		)
+		assert.NotNil(t, f)
+	})
 }
 
 func TestNewReportingPlugin(t *testing.T) {
-	mp := ktypes.NewMockPerformLogProvider(t)
-	hs := ktypes.NewMockHeadSubscriber(t)
-
-	f := &keepersReportingFactory{
-		registry:       ktypes.NewMockRegistry(t),
-		encoder:        ktypes.NewMockReportEncoder(t),
-		headSubscriber: hs,
-		perfLogs:       mp,
-		logger:         log.New(io.Discard, "test", 0),
-		config: ReportingFactoryConfig{
-			CacheExpiration:       30 * time.Second,
-			CacheEvictionInterval: 5 * time.Second,
-			MaxServiceWorkers:     1,
-			ServiceQueueLength:    10,
-		},
-	}
-
-	mp.Mock.On("PerformLogs", mock.Anything).
-		Return([]ktypes.PerformLog{}, nil).
-		Maybe()
-	mp.Mock.On("StaleReportLogs", mock.Anything).
-		Return([]ktypes.StaleReportLog{}, nil).
-		Maybe()
-
-	digest := [32]byte{}
-	digestStr := fmt.Sprintf("%32s", "test")
-	copy(digest[:], []byte(digestStr)[:32])
-
-	offchainConfig, err := json.Marshal(ktypes.OffchainConfig{
-		GasLimitPerReport:    500000,
-		GasOverheadPerUpkeep: 300000,
+	t.Run("fails to decode the offchain config", func(t *testing.T) {
+		mp := mocks.NewPerformLogProvider(t)
+		hs := mocks.NewHeadSubscriber(t)
+
+		f := &keepersReportingFactory{
```

Comment:
some of these logics are repeated multiple times. is there a way to consolidate them?

Reply:
There is for sure, but I actually try to keep code reuse to a minimum in tests; I think as business logic changes over time, the reused code in existing tests becomes brittle and it either becomes difficult to factor in new testing requirements to the reused functions, or you have to make more extensive changes to the reused code, which can impact all the tests that call that reused code. I think it's nice when all of a test's dependencies are inlined in the test itself, even if there is repetition. But if you feel strongly the other way I can totally extract the repeated blocks into a common function 👍